src/hotspot/share/opto/callnode.cpp

```diff
@@ -1665,6 +1669,276 @@ Node *AllocateArrayNode::make_ideal_length(const TypeOopPtr* oop_type, PhaseTran
+        input = may_be_allocate->as_Allocate()->result_cast();
+        assert(input->is_CheckCastPP(), "input to phi is not checkcastpp");
+      }
+    }
+
+    init_req(i, input);
+  }
+
+  initialize_memory_edges(C, igvn);
+
+  this->raise_bottom_type(ram_t);
+  igvn->set_type(this, ram_t);
+
+  C->add_macro_node(this);
+}
+
+void ReducedAllocationMergeNode::initialize_memory_edges(Compile* C, PhaseIterGVN* igvn) {
+  Node* region            = this->in(0);
+  ciInstanceKlass* iklass = _klass->as_instance_klass();
+  int nfields             = iklass->nof_nonstatic_fields();
+
+  // Make sure we have an entry for each base+field combination
+  register_offset_of_all_fields(NULL);
+
+  // Search for a memory edge matching base+field alias_index
+  for (uint i = 1, matches = 0; i <= _number_of_bases; i++) {
+    Node* base = this->in(i);
+    const TypeOopPtr *base_t = igvn->type(base)->isa_oopptr();
+
+    if (base_t != NULL) {
+      int base_offset = base_idx(base, 0);
+
+      for (int j = 0; j < nfields; j++) {
+        ciField* field          = iklass->nonstatic_field_at(j);
+        int offset              = field->offset();
+        const TypeOopPtr *tinst = base_t->add_offset(offset)->isa_oopptr();
+        const int alias_idx     = C->get_alias_index(tinst);
+        const int fields_offset = field_idx(offset);
+
+        for (DUIterator_Fast imax, i = region->fast_outs(imax); i < imax; i++) {
+          Node* memory = region->fast_out(i);
+          if (memory->is_Phi() && C->get_alias_index(memory->adr_type()) == alias_idx) {
+            set_req(fields_offset + base_offset, memory);
+          }
+        }
+      }
+    }
+  }
+
+  // Try to find a BOT memory Phi coming from same region
+  for (DUIterator_Fast imax, i = region->fast_outs(imax); i < imax; i++) {
+    Node* n = region->fast_out(i);
+    if (n->is_Phi() && n->bottom_type() == Type::MEMORY) {
+      if (C->get_alias_index(n->adr_type()) == Compile::AliasIdxBot) {
+        register_offset_of_all_fields(n);
+        break;
+      }
+    }
+  }
+}
+
+void ReducedAllocationMergeNode::register_offset_of_all_fields(Node* memory) {
+  ciInstanceKlass* iklass = _klass->as_instance_klass();
+  int nfields             = iklass->nof_nonstatic_fields();
+
+  for (int j = 0; j < nfields; j++) {
+    ciField* field = iklass->nonstatic_field_at(j);
+    int offset = field->offset();
+
+    register_offset(offset, memory);
+  }
+}
+
+void ReducedAllocationMergeNode::register_offset(int offset, Node* memory) {
+  assert(offset > 0, "Offset of use should be >= 0.");
+
+  if ((*_fields_and_memories)[(void*)(intptr_t)offset] == NULL) {
+    _fields_and_memories->Insert((void*)(intptr_t)offset, (void*)(intptr_t)req());
+
+    for (uint b_idx = 1; b_idx <= _number_of_bases; ++b_idx) {
+      add_req( (memory != NULL && memory->is_Phi() && memory->in(0) == in(0)) ?
+                    memory->in(b_idx) :
+                    memory);
+    }
+  }
+  else {
```

Comment:
Style - should be one line: ` } else {`.
I noticed that here and in other places in new code you put `else` and `else if (` on separate line from `}`. We use one line in such cases.